README.md
# blake2b

A dart implement of  blake2b referenced from [blacknet](https://blacknet.ninja/) and it's [source code](https://gitlab.com/blacknet-ninja/blacknet).

## Getting Started

```dart
///
/// CASE 1 - hash utf8 string
///

// what we expect
// bip = mean fury common entire zoo cash fragile dilemma retire appear insect park
// hash = A064E3C0E7D606F7EE717718FD23B943328ACD8CC20DE29B4053FC1C9D3888F6
// bytes = -96 100 -29 -64 -25 -42 6 -9 -18 113 119 24 -3 35 -71 67 50 -118 -51 -116 -62 13 -30 -101 64 83 -4 28 -99 56 -120 -10

Int8List expect = Int8List.fromList([-96, 100, -29, -64, -25, -42, 6, -9, -18, 113, 119, 24, -3, 35, -71, 67, 50, -118, -51, -116, -62, 13, -30, -101, 64, 83, -4, 28, -99, 56, -120, -10]);
Uint8List uexpect = Utils.int8list2uint8list(expect);

var bip = "mean fury common entire zoo cash fragile dilemma retire appear insect park";
var upperCaseHash = "A064E3C0E7D606F7EE717718FD23B943328ACD8CC20DE29B4053FC1C9D3888F6";

Int8List message = Int8List(bip.length);
for (int i = 0; i < bip.length; i++) {
    message[i] = bip.codeUnitAt(i);
}

// hash from bytes to bytes
var umessage = Utils.int8list2uint8list(message);

var bytes = Blake2bHash.hash(umessage, 0, umessage.length);
assert(_uint8ListEqual(bytes, uexpect));

// hash from hex string to hex string
var lowerCaseHash = Blake2bHash.hashHexString2HexString(HEX.encode(message));
assert(lowerCaseHash == upperCaseHash.toLowerCase());

// hash from utf8 string to hex string
lowerCaseHash = Blake2bHash.hashUtf8String2HexString(bip);
assert(lowerCaseHash == upperCaseHash.toLowerCase());

///
/// CASE 2 - hash bytes
///

// another test
// hexString = 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000eea2f40ef9c40caffc91d82c3332456039d9562290fe0957bad9ca7bc8e9159900000014d7338a9fc7b42de9c2045635a2bf2ec628f47adec76f8ee6caaa5397025b74e400000000000186a000ae0000000000989680a54407d81a2e4d5c72b8d784a56f88736c64bdf90883c12b6ffff6ec892b2b0c00846d656d6f
// hash = ff773f774a4489c98429d37ba1e8c91ce0ac26af27cd32c1b6ef17843e87eb39
var hexString =
    "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000eea2f40ef9c40caffc91d82c3332456039d9562290fe0957bad9ca7bc8e9159900000014d7338a9fc7b42de9c2045635a2bf2ec628f47adec76f8ee6caaa5397025b74e400000000000186a000ae0000000000989680a54407d81a2e4d5c72b8d784a56f88736c64bdf90883c12b6ffff6ec892b2b0c00846d656d6f";
var messageBytes = Utils.int8list2uint8list(Int8List.fromList(HEX.decode(hexString)));
var hash = Blake2bHash.hash(messageBytes, 64, messageBytes.length - 64);
var expectedHash = Utils.int8list2uint8list(Int8List.fromList(HEX.decode("ff773f774a4489c98429d37ba1e8c91ce0ac26af27cd32c1b6ef17843e87eb39")));
assert(_uint8ListEqual(hash, expectedHash));

///
/// CASE 3 - custom
///

String preSalt = 'preSalt';
int digestSize = 16;
Uint8List bPreSalt = Uint8List.fromList(preSalt.codeUnits);
Uint8List salt = Uint8List(digestSize);
final Blake2b blake2b = Blake2b.blake2b(null, digestSize, null, null);
blake2b.update(bPreSalt, 0, bPreSalt.length);
Uint8List hashBytes = blake2b.digest(salt, 0);
var hashStr = HEX.encode(hashBytes);
print(hashStr);
```

## pub ref

+ [ref](https://dart.cn/tools/pub/publishing)


CHANGELOG.md
## [0.0.1] - basic implement of blake2b

* just follow the blacknet's blake2b to implment it with dart.

## [0.0.2] - wrap a Blake2bHash for common use

## [0.1.0] - use Uint8List instead of Int8List

## [0.1.2] - fix bug due to a wrong implement of array copy

## [0.2.1] - null safety, remove flutter dependency

## [0.2.2] - bugfix: bytes2long() overflow when in chrome see [Issue 1](https://github.com/riclava/blake2b/issues/1)

## [0.2.3] - feature: new constructor `Blake2b.blake2b(final Uint8List? key, final int digestSize, final Uint8List? salt, final Uint8List? personalization)`


test/blake2b_test.dart
///
/// Copyright (c) [2019] [riclava]
/// [blake2b] is licensed under the Mulan PSL v1.
/// You can use this software according to the terms and conditions of the Mulan PSL v1.
/// You may obtain a copy of Mulan PSL v1 at:
///     http://license.coscl.org.cn/MulanPSL
/// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
/// IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
/// PURPOSE.
/// See the Mulan PSL v1 for more details.
///
import 'dart:typed_data';

import 'package:blake2b/blake2b.dart';
import 'package:blake2b/blake2b_hash.dart';
import 'package:blake2b/utils.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:hex/hex.dart';

void main() {
  bool _uint8ListEqual(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (int i = 0; i < a.length; i++) {
      if (a[i] != b[i]) {
        return false;
      }
    }
    return true;
  }

  test('test', () {
    ///
    /// CASE 1 - hash utf8 string
    ///

    // what we expect
    // bip = mean fury common entire zoo cash fragile dilemma retire appear insect park
    // hash = A064E3C0E7D606F7EE717718FD23B943328ACD8CC20DE29B4053FC1C9D3888F6
    // bytes = -96 100 -29 -64 -25 -42 6 -9 -18 113 119 24 -3 35 -71 67 50 -118 -51 -116 -62 13 -30 -101 64 83 -4 28 -99 56 -120 -10

    Int8List expect = Int8List.fromList([-96, 100, -29, -64, -25, -42, 6, -9, -18, 113, 119, 24, -3, 35, -71, 67, 50, -118, -51, -116, -62, 13, -30, -101, 64, 83, -4, 28, -99, 56, -120, -10]);
    Uint8List uexpect = Utils.int8list2uint8list(expect);

    var bip = "mean fury common entire zoo cash fragile dilemma retire appear insect park";
    var upperCaseHash = "A064E3C0E7D606F7EE717718FD23B943328ACD8CC20DE29B4053FC1C9D3888F6";

    Int8List message = Int8List(bip.length);
    for (int i = 0; i < bip.length; i++) {
      message[i] = bip.codeUnitAt(i);
    }

    // hash from bytes to bytes
    var umessage = Utils.int8list2uint8list(message);

    var bytes = Blake2bHash.hash(umessage, 0, umessage.length);
    assert(_uint8ListEqual(bytes, uexpect));

    // hash from hex string to hex string
    var lowerCaseHash = Blake2bHash.hashHexString2HexString(HEX.encode(message));
    assert(lowerCaseHash == upperCaseHash.toLowerCase());

    // hash from utf8 string to hex string
    lowerCaseHash = Blake2bHash.hashUtf8String2HexString(bip);
    assert(lowerCaseHash == upperCaseHash.toLowerCase());

    ///
    /// CASE 2 - hash bytes
    ///

    // another test
    // hexString = 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000eea2f40ef9c40caffc91d82c3332456039d9562290fe0957bad9ca7bc8e9159900000014d7338a9fc7b42de9c2045635a2bf2ec628f47adec76f8ee6caaa5397025b74e400000000000186a000ae0000000000989680a54407d81a2e4d5c72b8d784a56f88736c64bdf90883c12b6ffff6ec892b2b0c00846d656d6f
    // hash = ff773f774a4489c98429d37ba1e8c91ce0ac26af27cd32c1b6ef17843e87eb39
    var hexString =
        "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000eea2f40ef9c40caffc91d82c3332456039d9562290fe0957bad9ca7bc8e9159900000014d7338a9fc7b42de9c2045635a2bf2ec628f47adec76f8ee6caaa5397025b74e400000000000186a000ae0000000000989680a54407d81a2e4d5c72b8d784a56f88736c64bdf90883c12b6ffff6ec892b2b0c00846d656d6f";
    var messageBytes = Utils.int8list2uint8list(Int8List.fromList(HEX.decode(hexString)));
    var hash = Blake2bHash.hash(messageBytes, 64, messageBytes.length - 64);
    var expectedHash = Utils.int8list2uint8list(Int8List.fromList(HEX.decode("ff773f774a4489c98429d37ba1e8c91ce0ac26af27cd32c1b6ef17843e87eb39")));
    assert(_uint8ListEqual(hash, expectedHash));

    ///
    /// CASE 3 - custom
    ///

    String preSalt = 'preSalt';
    int digestSize = 16;
    Uint8List bPreSalt = Uint8List.fromList(preSalt.codeUnits);
    Uint8List salt = Uint8List(digestSize);
    final Blake2b blake2b = Blake2b.blake2b(null, digestSize, null, null);
    blake2b.update(bPreSalt, 0, bPreSalt.length);
    Uint8List hashBytes = blake2b.digest(salt, 0);
    var hashStr = HEX.encode(hashBytes);
    print(hashStr);
  });
}


lib/hex.dart
///
/// Copyright (c) [2019] [name of copyright holder]
/// [Software Name] is licensed under the Mulan PSL v1.
/// You can use this software according to the terms and conditions of the Mulan PSL v1.
/// You may obtain a copy of Mulan PSL v1 at:
///     http://license.coscl.org.cn/MulanPSL
/// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
/// IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
/// PURPOSE.
/// See the Mulan PSL v1 for more details.
///
import 'dart:typed_data';

import 'package:blake2b/ushr.dart';

class Hex {
  static final Int8List _rawHexTable = Int8List.fromList([
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    10,
    11,
    12,
    13,
    14,
    15,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    10,
    11,
    12,
    13,
    14,
    15
  ]);

  static final List<String> _lowerCaseChars = [
    '0',
    '1',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    'a',
    'b',
    'c',
    'd',
    'e',
    'f'
  ];
  static final List<String> _upperCaseChars = [
    '0',
    '1',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    'A',
    'B',
    'C',
    'D',
    'E',
    'F'
  ];

  static Int8List decodeHex(String string) {
    var length = string.length;
    if ((length & 0x01) != 0) {
      return null; // Odd number of characters.
    }

    List<int> result = [];

    for (int i = 0; i < length; i += 2) {
      var firstIndex = _hexTable(string.codeUnitAt(i));
      var secondIndex = _hexTable(string.codeUnitAt(i + 1));

      if (firstIndex == -1 || secondIndex == -1) return null;

      var octet = (firstIndex << 4) | secondIndex;
      result.add(octet.toInt());
    }
    return Int8List.fromList(result);
  }

  static String encodeHexString(Int8List l, bool lowerCase) {
    List<String> table = lowerCase ? _lowerCaseChars : _upperCaseChars;
    String result = "";
    for (int i = 0; i < l.length; i++) {
      var octet = l[i];
      int firstIndex = Ushr.ushr((octet & 0xF0), 4);
      var secondIndex = octet & 0x0F;
      result += table[firstIndex];
      result += table[secondIndex];
    }
    return result;
  }

  static int _hexTable(int elem) {
    if (elem > 0 && elem < _rawHexTable.length) {
      return _rawHexTable[elem];
    }
    return -1;
  }
}


lib/hash.dart
///
/// Copyright (c) [2019] [name of copyright holder]
/// [Software Name] is licensed under the Mulan PSL v1.
/// You can use this software according to the terms and conditions of the Mulan PSL v1.
/// You may obtain a copy of Mulan PSL v1 at:
///     http://license.coscl.org.cn/MulanPSL
/// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
/// IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
/// PURPOSE.
/// See the Mulan PSL v1 for more details.
///
import 'dart:typed_data';

import 'package:blake2b/blake2b.dart';
import 'package:fixnum/fixnum.dart';

class Hash {

  static final int _size = 32;
  
  static final int _digest_size = _size * 8;

  List<Int64> bytes;

  utf8(String string) {
    // TODO implement me
  }

  static Int8List hash(Int8List message, int offset, int len) {
    Int8List bytes = Int8List(_size);
    var b = Blake2b(_digest_size);
    b.update(message, offset, len);
    b.digest(bytes, 0);
    return bytes;
  }
}

lib/const.dart
///
/// Copyright (c) [2019] [name of copyright holder]
/// [Software Name] is licensed under the Mulan PSL v1.
/// You can use this software according to the terms and conditions of the Mulan PSL v1.
/// You may obtain a copy of Mulan PSL v1 at:
///     http://license.coscl.org.cn/MulanPSL
/// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
/// IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
/// PURPOSE.
/// See the Mulan PSL v1 for more details.
///
import 'package:fixnum/fixnum.dart';

class Const {
  static final Int64 f0 = Int64.fromInts(0xffffffff, 0xffffffff);

  // Blake2b Initialization Vector:
  // Produced from the square root of primes 2, 3, 5, 7, 11, 13, 17, 19.
  // The same as SHA-512 IV.
  static final List<Int64> blake2bIv = [
    Int64.fromInts(0x6a09e667, 0xf3bcc908),
    Int64.fromInts(0xbb67ae85, 0x84caa73b),
    Int64.fromInts(0x3c6ef372, 0xfe94f82b),
    Int64.fromInts(0xa54ff53a, 0x5f1d36f1),
    Int64.fromInts(0x510e527f, 0xade682d1),
    Int64.fromInts(0x9b05688c, 0x2b3e6c1f),
    Int64.fromInts(0x1f83d9ab, 0xfb41bd6b),
    Int64.fromInts(0x5be0cd19, 0x137e2179),
  ];

  // Message word permutations:
  static final List<List<int>> blake2Sigma = [
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
    [11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4],
    [7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8],
    [9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13],
    [2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9],
    [12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11],
    [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
    [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
    [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0],
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3]
  ];

  static final int blockLengthBytes = 128; // bytes

  static final int rounds = 12; // to use for Catenas H'
}


lib/byte_utils.dart
///
/// Copyright (c) [2019] [riclava]
/// [blake2b] is licensed under the Mulan PSL v1.
/// You can use this software according to the terms and conditions of the Mulan PSL v1.
/// You may obtain a copy of Mulan PSL v1 at:
///     http://license.coscl.org.cn/MulanPSL
/// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
/// IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
/// PURPOSE.
/// See the Mulan PSL v1 for more details.
///
import 'dart:typed_data';

import 'package:fixnum/fixnum.dart';

class ByteUtils {
  static Int64 bytes2long(final Uint8List byteArray, final int offset) {
    // & 0xFF is useless
    return ((Int64.fromInts(0, byteArray[offset] & 0xFF)) |
        (Int64.parseInt((byteArray[offset + 1] & 0xFF).toString()) << 8) |
        (Int64.parseInt((byteArray[offset + 2] & 0xFF).toString()) << 16) |
        (Int64.parseInt((byteArray[offset + 3] & 0xFF).toString()) << 24) |
        (Int64.parseInt((byteArray[offset + 4] & 0xFF).toString()) << 32) |
        (Int64.parseInt((byteArray[offset + 5] & 0xFF).toString()) << 40) |
        (Int64.parseInt((byteArray[offset + 6] & 0xFF).toString()) << 48) |
        (Int64.parseInt((byteArray[offset + 7] & 0xFF).toString()) << 56));
  }

  static Uint8List long2bytes(final Int64 longValue) {
    return Uint8List.fromList(longValue.toBytes());
  }

  static Int64 rotr64(final Int64 x, final int rot) {
    return x.shiftRightUnsigned(rot) | x << (64 - rot);
  }
}


lib/blake2b_hash.dart
///
/// Copyright (c) [2019] [riclava]
/// [blake2b] is licensed under the Mulan PSL v1.
/// You can use this software according to the terms and conditions of the Mulan PSL v1.
/// You may obtain a copy of Mulan PSL v1 at:
///     http://license.coscl.org.cn/MulanPSL
/// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
/// IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
/// PURPOSE.
/// See the Mulan PSL v1 for more details.
///
import 'dart:typed_data';

import 'package:blake2b/blake2b.dart';
import 'package:hex/hex.dart';

class Blake2bHash {
  static final int _size = 32;

  static final int _digestSize = _size * 8;

  // hash from hex string to bytes
  static Uint8List hashHexString(String hex) {

    var bytes = HEX.decode(hex);
    return hash(bytes as Uint8List, 0, bytes.length);
  }

  // hash from hex string to hex string
  static String hashHexString2HexString(String hex) {
    return HEX.encode(hashHexString(hex));
  }

  // hash from utf8 string to bytes
  static Uint8List hashUtf8String(String string) {
    Uint8List bytes = Uint8List(string.length);
    for (int i = 0; i < string.length; i++) {
      bytes[i] = string.codeUnitAt(i);
    }
    return hash(bytes, 0, bytes.length);
  }

  // hash from utf8 string to hex string
  static String hashUtf8String2HexString(String string) {
    return HEX.encode(hashUtf8String(string));
  }

  static Uint8List hashWithDigestSize(int digestSize, Uint8List message) {
    Uint8List bytes = Uint8List(digestSize ~/ 8);
    var b = Blake2b(digestSize);
    b.update(message, 0, message.length);
    return b.digest(bytes, 0);
  }

  static Uint8List hash(Uint8List message, int offset, int len) {
    Uint8List bytes = Uint8List(_size);
    var b = Blake2b(_digestSize);
    b.update(message, offset, len);    
    return b.digest(bytes, 0);
  }
}


lib/blake2b.dart
///
/// Copyright (c) [2019] [riclava]
/// [blake2b] is licensed under the Mulan PSL v1.
/// You can use this software according to the terms and conditions of the Mulan PSL v1.
/// You may obtain a copy of Mulan PSL v1 at:
///     http://license.coscl.org.cn/MulanPSL
/// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
/// IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
/// PURPOSE.
/// See the Mulan PSL v1 for more details.
///
library blake2b;

import 'dart:typed_data';

import 'package:blake2b/byte_utils.dart';
import 'package:blake2b/const.dart';
import 'package:fixnum/fixnum.dart';

/// Blake2b
class Blake2b {
  Blake2b(int digestSize) {
    assert(digestSize == 128 || digestSize == 160 || digestSize == 256 || digestSize == 384 || digestSize == 512);

    this._buffer = Uint8List(Const.blockLengthBytes);
    this._keyLength = 0;
    this._digestSize = digestSize ~/ 8;
    init();
  }

  blake2bWithKey(final Uint8List key) {
    assert(key.length <= 64);
    this._buffer = Uint8List(Const.blockLengthBytes);

    this._key!.addAll(key);
    this._keyLength = key.length;

    this._buffer.addAll(key);
    this._bufferPos = Const.blockLengthBytes;
    _digestSize = 64;
    init();
  }

  Blake2b.blake2b(final Uint8List? key, final int digestSize, final Uint8List? salt, final Uint8List? personalization) {
    this._buffer = Uint8List(Const.blockLengthBytes);
    assert(1 <= digestSize && digestSize <= 64);
    this._digestSize = digestSize;

    if (null != salt) {
      assert(salt.length == 16);
      this._salt!.addAll(salt);
    }

    if (null != personalization) {
      assert(personalization.length == 16);
      this._personalization!.addAll(personalization);
    }

    if (null != key) {
      this._key!.addAll(key);
      this._keyLength = key.length;

      this._buffer.addAll(key);
      this._bufferPos = Const.blockLengthBytes; // zero padding
    }

    init();
  }

  blake2b(final Uint8List key, final int digestSize, final Uint8List salt, final Uint8List personalization) {
    this._buffer = Uint8List(Const.blockLengthBytes);
    assert(1 <= digestSize && digestSize <= 64);
    this._digestSize = digestSize;

    assert(salt.length == 16);
    this._salt!.addAll(salt);

    assert(personalization.length == 16);
    this._personalization!.addAll(personalization);

    this._key!.addAll(key);
    this._keyLength = key.length;

    this._buffer.addAll(key);
    this._bufferPos = Const.blockLengthBytes;

    init();
  }

  // General parameters
  int _digestSize = 64; // 1- 64 bytes
  int _keyLength = 0; // 0 - 64 bytes for keyed hashing for MAC
  Uint8List? _salt; // new byte[16];
  Uint8List? _personalization; // new byte[16];

  // The key
  Uint8List? _key;

  // whenever this buffer overflows, it will be processed
  // in the compress() function.
  // For performance issues, Int64 messages will not use this buffer.
  Uint8List _buffer = Uint8List(Const.blockLengthBytes); // new byte[BLOCK_LENGTH_BYTES];

  // Position of last inserted byte:
  int _bufferPos = 0; // a value from 0 up to 128

  List<Int64> _internalState = List<Int64>.filled(16, Int64.fromInts(0, 0)); // In the Blake2b paper it is called: v

  List<Int64>? _chainValue; // state vector, in the Blake2b paper it is called: h

  Int64 _t0 = Int64.fromInts(0, 0); // holds last significant bits, counter (counts bytes)
  Int64 _t1 = Int64.fromInts(0, 0); // counter: Length up to 2^128 are supported
  Int64 _f0 = Int64.fromInts(0, 0); // finalization flag, for last block: ~0L

  void init() {
    if (null == _chainValue) {
      final List<Int64> newChainValue = [];

      // 0
      newChainValue.add(Const.blake2bIv[0] ^ (_digestSize | (_keyLength << 8) | 0x1010000));
      // 0x1010000 = ((fanout << 16) | (depth << 24) | (leafLength << 32));
      // with fanout = 1; depth = 0; leafLength = 0;

      // 1
      newChainValue.add(Const.blake2bIv[1]); // ^ nodeOffset; with nodeOffset = 0;

      // 2
      newChainValue.add(Const.blake2bIv[2]); // ^ ( nodeDepth | (innerHashLength << 8) );
      // with nodeDepth = 0; innerHashLength = 0;

      // 3
      newChainValue.add(Const.blake2bIv[3]);

      // 4, 5
      newChainValue.add(Const.blake2bIv[4]);
      newChainValue.add(Const.blake2bIv[5]);

      if (null != _salt) {
        newChainValue[4] ^= (ByteUtils.bytes2long(_salt!, 0));
        newChainValue[5] ^= (ByteUtils.bytes2long(_salt!, 8));
      }

      // 6, 7
      newChainValue.add(Const.blake2bIv[6]);
      newChainValue.add(Const.blake2bIv[7]);

      if (null != _personalization) {
        newChainValue[6] ^= (ByteUtils.bytes2long(_personalization!, 0));
        newChainValue[7] ^= (ByteUtils.bytes2long(_personalization!, 8));
      }

      _chainValue = newChainValue;
    }
  }

  void _initializeInternalState() {
    _internalState = _arrayCopyI64(_chainValue, 0, _internalState, 0, _chainValue!.length);
    _internalState = _arrayCopyI64(Const.blake2bIv, 0, _internalState, _chainValue!.length, 4);

    _internalState[12] = _t0 ^ Const.blake2bIv[4];
    _internalState[13] = _t1 ^ Const.blake2bIv[5];
    _internalState[14] = _f0 ^ Const.blake2bIv[6];
    _internalState[15] = Const.blake2bIv[7];
  }

  void reset() {
    _bufferPos = 0;
    _f0 = Int64.fromInts(0, 0);
    _t0 = Int64.fromInts(0, 0);
    _t1 = Int64.fromInts(0, 0);
    _chainValue = null;

    _buffer.fillRange(0, _buffer.length, 0);

    if (_key != null) {
      _buffer.addAll(_key!);
      _bufferPos = Const.blockLengthBytes; // zero padding
    }
    init();
  }

  void clearKey() {
    if (null != _key) {
      _key!.fillRange(0, _key!.length, 0);
      _buffer.fillRange(0, _buffer.length, 0);
    }
  }

  void clearSalt() {
    if (null != _salt) {
      _salt!.fillRange(0, _salt!.length, 0);
    }
  }

  void update0(final int b) {
    final int remainingLength = Const.blockLengthBytes - _bufferPos;
    if (remainingLength == 0) {
      _t0 += Const.blockLengthBytes;
      if (_t0 == 0) {
        _t1++;
      }
      _compress(_buffer, 0);
      _buffer.fillRange(0, _buffer.length, 0);
      _buffer[0] = b;
      _bufferPos = 1;
    } else {
      _buffer[_bufferPos] = b;
      _bufferPos++;
    }
  }

  void update(Uint8List message, int offset, int len) {
    if (0 == len) {
      return;
    }

    int remainingLength = 0;

    if (0 != _bufferPos) {
      // commenced, incomplete buffer

      // complete the buffer:
      remainingLength = Const.blockLengthBytes - _bufferPos;
      if (remainingLength < len) {
        // full buffer + at least 1 byte
        _buffer = _arrayCopy(message, offset, _buffer, _bufferPos, remainingLength);
        _t0 += Const.blockLengthBytes;
        if (_t0 == 0) {
          // if message > 2^64
          _t1++;
        }
        _compress(_buffer, 0);
        _bufferPos = 0;
        _buffer.fillRange(0, _buffer.length, 0);
      } else {
        _buffer = _arrayCopy(message, offset, _buffer, _bufferPos, len);
        _bufferPos += len;
        return;
      }
    }

    // process blocks except last block (also if last block is full)
    int messagePos;
    final int blockWiseLastPos = offset + len - Const.blockLengthBytes;

    // block wise 128 bytes
    for (messagePos = offset + remainingLength; messagePos < blockWiseLastPos; messagePos += Const.blockLengthBytes) {
      // without buffer:
      _t0 += Const.blockLengthBytes;

      if (_t0 == 0) {
        _t1++;
      }

      _compress(message, messagePos);
    }

    // fill the buffer with left bytes, this might be a full block
    _buffer = _arrayCopy(message, messagePos, _buffer, 0, offset + len - messagePos);

    _bufferPos += (offset + len) - messagePos;
  }

  Uint8List digest(Uint8List out, final int outOffset) {
    _f0 = Const.f0;
    _t0 += _bufferPos;

    if (0 < _bufferPos && _t0 == 0) {
      _t1++;
    }

    _compress(_buffer, 0);

    _buffer.fillRange(0, _buffer.length, 0);
    _internalState = List<Int64>.filled(_internalState.length, Int64.fromInts(0, 0));

    for (int i = 0; i < _chainValue!.length && (i * 8 < _digestSize); i++) {
      Uint8List bytes = ByteUtils.long2bytes(_chainValue![i]);

      if ((i * 8) < (_digestSize - 8)) {
        out = _arrayCopy(bytes, 0, out, outOffset + (i * 8), 8);
      } else {
        out = _arrayCopy(bytes, 0, out, outOffset + (i * 8), _digestSize - (i * 8));
      }
    }

    _chainValue = List<Int64>.filled(_chainValue!.length, Int64.fromInts(0, 0));

    reset();

    return out;
  }

  void _compress(Uint8List message, int messagePos) {
    _initializeInternalState();

    List<Int64> m = List<Int64>.filled(16, Int64.fromInts(0, 0));
    for (int j = 0; j < 16; j++) {
      m[j] = ByteUtils.bytes2long(message, messagePos + j * 8);
    }

    for (int round = 0; round < Const.rounds; round++) {
      _g(m[Const.blake2Sigma[round][0]], m[Const.blake2Sigma[round][1]], 0, 4, 8, 12);
      _g(m[Const.blake2Sigma[round][2]], m[Const.blake2Sigma[round][3]], 1, 5, 9, 13);
      _g(m[Const.blake2Sigma[round][4]], m[Const.blake2Sigma[round][5]], 2, 6, 10, 14);
      _g(m[Const.blake2Sigma[round][6]], m[Const.blake2Sigma[round][7]], 3, 7, 11, 15);
      _g(m[Const.blake2Sigma[round][8]], m[Const.blake2Sigma[round][9]], 0, 5, 10, 15);
      _g(m[Const.blake2Sigma[round][10]], m[Const.blake2Sigma[round][11]], 1, 6, 11, 12);
      _g(m[Const.blake2Sigma[round][12]], m[Const.blake2Sigma[round][13]], 2, 7, 8, 13);
      _g(m[Const.blake2Sigma[round][14]], m[Const.blake2Sigma[round][15]], 3, 4, 9, 14);
    }

    for (int offset = 0; offset < _chainValue!.length; offset++) {
      _chainValue![offset] = _chainValue![offset] ^ _internalState[offset] ^ _internalState[offset + 8];
    }
  }

  void _g(Int64 m1, Int64 m2, int posA, int posB, int posC, int posD) {
    _internalState[posA] = _internalState[posA] + _internalState[posB] + m1;
    _internalState[posD] = ByteUtils.rotr64(_internalState[posD] ^ _internalState[posA], 32);

    _internalState[posC] = _internalState[posC] + _internalState[posD];
    _internalState[posB] = ByteUtils.rotr64(_internalState[posB] ^ _internalState[posC], 24); // replaces 25 of BLAKE

    _internalState[posA] = _internalState[posA] + _internalState[posB] + m2;
    _internalState[posD] = ByteUtils.rotr64(_internalState[posD] ^ _internalState[posA], 16);

    _internalState[posC] = _internalState[posC] + _internalState[posD];
    _internalState[posB] = ByteUtils.rotr64(_internalState[posB] ^ _internalState[posC], 63);
  }

  Uint8List _arrayCopy(Uint8List src, int srcOffset, Uint8List dst, int dstOffset, int length) {
    for (var i = 0; i < length; i++) {
      dst[dstOffset + i] = src[srcOffset + i];
    }
    return dst;
  }

  List<Int64> _arrayCopyI64(List<Int64>? src, int srcOffset, List<Int64> dst, int dstOffset, int length) {
    for (var i = 0; i < length; i++) {
      dst[dstOffset + i] = src![srcOffset + i];
    }
    return dst;
  }
}
